Start worker processes from leader

tsyesika opened the pull request:
This commit is part of the work on #901 - It will be merged onto `lwaftr-multiproc` where the work will take place. Once #901 has been fully implemented the branch will be merged into `lwaftr` in a separate PR.

This moves creation and management of the follower processes to the leader. This is done by passing in a new parameter (instead of follower PIDs) called `worker_start_fn` which will start a worker when called and return it's PID. There is also a change to the `setup_fn`, it now returns a list of followers it wants starting for a given configuration. Each follower is given an ID (in the case of the lwAFTR it's the instance key, the PCI device). For each follower the leader maintains a app graph.

From this configuration multiple lwaftr processes will be run if they are provided. This does not handle configuration changes to modify the instance list.

Commit:
Start worker processes from leader

This commit changes the setup of the lwAFTR to run the worker
processes from the Leader process instead of from the setup
itself.

There is a new argument for the leader which is the function to start
the worker (follower). The worker function needs to setup the follower
and return the PID of it.

There is also a change to the setup function that requires it to
return a table of ID to app graph. The setup function should return
the same ID for the same function (i.e. it should not be random but
consistantly unique to itself).

Commit:
Cleanup setup.lua to remove produce_instance_configs calls

This removes the unnecessary calls to produce an instance
configuration as that's already being passed to them. It also fixes a
bug where they were all being passed the same instances' configs.

Commit:
Fixes Leader update configuration

This adds multiprocess support to the leader update configuration
code.

wingo commented on src/apps/config/leader.lua:
```diff
@@ -27,7 +27,7 @@ Leader = {
       initial_configuration = {required=true},
       schema_name = {required=true},
-      follower_pids = {required=true},
+      worker_start_fn = {required=true},
```
What do you think about replacing this with the Lua code that would be run in the follower?  That way the leader handlers the "worker.start" bit.

wingo commented on src/apps/config/leader.lua:
```diff
@@ -70,16 +68,25 @@ end
+      local in_place_dependencies =
+         self.support.update_mutable_objects_embedded_in_app_initargs (
+            app_graph.new(), config, self.schema_name, 'load', '/',
+            self.current_configuration)
```
Hum!  Here I'm not sure.  So the leader is exposing a single YANG tree, and then the setup function maps that onto N app graphs.  But when a user updates a part of that YANG tree, some subset of those N graphs will have been updated in place.  I think we need to build one "in_place_dependencies" for the whole system so we can know what to restart when a user updates the YANG configuration -- i.e. instead of having update_mutable_objects_embedded_in_app_initargs take a single app graph, probably it should take a table of all app graphs, keyed by follower ID.  WDYT?

wingo commented on src/apps/config/leader.lua:
```diff
@@ -418,21 +425,23 @@ end
-      self.support.update_mutable_objects_embedded_in_app_initargs (
-         self.current_in_place_dependencies, self.current_app_graph,
-         verb, path, ...)
+   local new_graph = update_fn(self.current_configuration, ...)
```
Is this right?  It looks like before, update_fn returned a configuration, whereas now you are having it return a graph.  I looked through the rest of the file for relevant changes but I must have missed them.

tsyesika replied:
This is a mistake, it's suppose to be a call to `self.setup_fn`

wingo commented on src/apps/config/leader.lua:
```diff
@@ -741,10 +750,10 @@ function selftest ()
    end
    app_graph.app(graph, "leader", Leader,
-                 {setup_fn=setup_fn, follower_pids={S.getpid()},
+                 {setup_fn=setup_fn, worker_start_fn=S.getpid,
```
Here we should definitely test spawning a worker process :)

wingo reviewed (commented):
Looking good, some questions :)

Commit:
Maintain a single in_place_dependencies

This change makes it so that the leader maintains a single set
of dependencies rather than each follower maintaining their own. To
do this the table backing each dependency also now lists a table of
follower IDs so that it's known which followers need to restart.

The main change to this is that the following support functions now
take a table with id and app graph rather than simply the app graph:

- update_mutable_objects_embedded_in_app_initargs
- compute_apps_to_restart_after_configuration_update

wingo commented on src/apps/config/leader.lua:
```diff
@@ -70,16 +69,31 @@ end
-   self.current_in_place_dependencies =
+
+   -- Start the followers and configure them.
+   local follower_configurations = self.setup_fn(configuration)
```
It would be nice if the word "configuration" applied either to the user-visible configuration of the network service, as instance data of the YANG schema, or to the Snabb-internal configuration of the graph of applications and links.  WDYT?  If you don't have a preference as to which is which, let's reserve "configuration" for the user-visible YANG config, and use the term "app graph" or "graph" to describe the Snabb-internal graph of apps and links.  Then I guess we would call this variable "worker_app_graphs" or something

wingo commented on src/apps/config/leader.lua:
```diff
@@ -70,16 +69,31 @@ end
-   local actions = self.support.compute_config_actions(
-      initial_app_graph, self.current_app_graph, {}, 'load')
-   self:enqueue_config_actions(actions)
+            app_graph.new(), follower_configurations, self.schema_name, 'load',
```
This doesn't look right to me; the default `update_mutable_objects_embedded_in_app_initargs` function goes like this:

```lua
   update_mutable_objects_embedded_in_app_initargs = function(
         in_place_dependencies, app_graph, schema_name, verb, path, arg)
      return compute_mutable_objects_embedded_in_app_initargs(app_graph)
   end,
```

Obviously this is generic fallback, but still.  I think the first argument should probably still be the initial `in_place_dependencies` which is the empty map `{}`, and then we need to replace the `app_graph` argument with a table of app graphs (`follower_configurations` or `worker_app_graphs` or whatever that ends up being).  Of course then we have to adapt the implementation to expect a table of app graphs, but I think you've done that below.  WDYT?

wingo commented on src/apps/config/leader.lua:
```diff
@@ -70,16 +69,31 @@ end
+   local follower_configurations = self.setup_fn(configuration)
+
+   -- Calculate the dependences
+   self.in_place_dependencies =
```
Why rename from `current_in_place_dependencies`?  What's the difference between this one and the one defined above?

wingo commented on src/apps/config/leader.lua:
```diff
@@ -421,18 +435,19 @@ function Leader:update_configuration (update_fn, verb, path, ...)
+   end
    self.current_configuration = new_config
-   self.current_in_place_dependencies =
+   self.in_place_dependencies =
```
For me I would leave the `current_` moniker to match the `current_configuration`, but I dunno...

wingo commented on src/apps/config/leader.lua:
```diff
@@ -421,18 +435,19 @@ function Leader:update_configuration (update_fn, verb, path, ...)
-      self.current_app_graph, new_app_graph, to_restart, verb, path, ...)
-   self:enqueue_config_actions(actions)
-   self.current_app_graph = new_app_graph
+   local new_graph = self.setup_fn(new_config, ...)
```
Perhaps rename to `new_graphs`, as it's a table of app graphs?

wingo commented on src/apps/config/leader.lua:
```diff
@@ -421,18 +435,19 @@ function Leader:update_configuration (update_fn, verb, path, ...)
-   self:enqueue_config_actions(actions)
-   self.current_app_graph = new_app_graph
+   local new_graph = self.setup_fn(new_config, ...)
+   for id, follower in pairs(self.followers) do
```
Probably here we should be noticing new followers in the `new_graphs`, starting them, and stopping old followers that are no longer configured.  WDYT?

tsyesika replied:
Sure. I was going to be doing that in a different PR but i can do the work and lump it in this one if you'd like.

wingo commented on src/apps/config/support.lua:
```diff
@@ -98,11 +98,20 @@ local function compute_objects_maybe_updated_in_place (schema_name, config,
       local tab = accum[obj]
-      if not tab then tab = {}; accum[obj] = tab end
-      table.insert(tab, name)
+      local debugit = true
```
What's this?

tsyesika replied:
Left over debug statement, clearly got through my pre-push check. I'll remove it :)

wingo commented on src/apps/config/support.lua:
```diff
@@ -119,12 +128,14 @@ local function record_mutable_objects_embedded_in_app_initarg (name, obj, accum)
--- list of app names.
-local function compute_mutable_objects_embedded_in_app_initargs (app_graph)
+-- Takes a table of follower ids (app_graph_map) and returns
+-- "in-place dependencies": a table mapping mutable object -> list of app names
```
I think we probably need to return a table like `{follower_id => {mutable_object => {app_name...}}}` or something.  Or, `{mutable_object=>{{follower=follower_id, app_names={...}}...}}`.  I mean, an app name is ambiguous -- it could be you only mean to refer to app A in follower M, not A in follower N.

wingo reviewed (commented):
Looking great, some comments.

Commit:
Fix nits in variable names and comments

wingo commented on src/apps/config/leader.lua:
```diff
@@ -741,25 +765,20 @@ function selftest ()
-   assert(app.app_table.source)
-   assert(app.app_table.sink)
-   assert(app.link_table["source.foo -> sink.bar"])
+   assert(app.app_table.leader.followers[1])
```
What's the `1` here for?  I would have thought that you'd get some other ID.  Maybe I am missing this :)

tsyesika replied:
The device id for the test leader app is `1`. This could be different if preferable. The `setup_fn` returns a table of `{id => app graph, ...}` in this case our `setup_fn` simply returns `{graph}` (https://github.com/Igalia/snabb/blob/c7b918c7d39c8758bd71e6b378c0fed797f50200/src/apps/config/leader.lua#L768) so lua will give it the ID of 1.

wingo replied:
Ah makes sense to me, tx for explanation.

wingo commented on src/apps/config/support.lua:
```diff
@@ -98,11 +98,18 @@ local function compute_objects_maybe_updated_in_place (schema_name, config,
 end
 
-local function record_mutable_objects_embedded_in_app_initarg (name, obj, accum)
+local function record_mutable_objects_embedded_in_app_initarg (id, name, obj, accum)
```
I am having a hard time figuring out what `id` and `name` refer to.  If it's `follower_id` and `app_name` then can we make that change?

wingo commented on src/program/lwaftr/bench/bench.lua:
```diff
@@ -58,19 +77,32 @@ function run(args)
+         timer.activate(timer.new('start_sampling', cycle, 1e9))
       end
-      timer.activate(timer.new('spawn_csv_stats', start_sampling, 10e6))
+      timer.activate(timer.new('start_sampling', cycle, 1e9))
```
Here you can add a fourth argument, 'repeating', to make a repeating timer.  But how do we want to treat statistics for multiple workers?  Sum or list separately?  A question for a followup, perhaps.

tsyesika replied:
I've changed it to be repeating.

I'm not sure if how to handle multiple sets of statistics. I'm currently just listing them separately, I don't know if that should change. It probably shouldn't print the header multiple times which it does. I can disable that for now and if they want to be summed or something other than just listed individually for each instance maybe that can come later.

wingo reviewed (approved):
LGTM, just some small comments.

Commit:
Improve variable name clarity and make timer repeating

This improves the variable names to better describe what they
represent. It also makes the timer in the bench.lua code
repeating rather than simply creating lots of new timers everytime
it's called.

Commit:
Move sampling to setup

This moves the finding of followers for the leader and starting of
sampling to the setup function so it can be shared across both the
bench command and the run command.

In addition to this it also prevents the header from being written
multiple times.

Commit:
Merge remote-tracking branch 'origin/lwaftr' into lwaftr-worker-in-leader

wingo commented:
still lgtm :)
